File: src/languages/japanese/japanese.ts
/**
 * Japanese (ja) phonemizer — Standard/Tokyo, canonical IPA.
 *   text → segmentText (bunsetsu spaces) → per run: applyReadings (kanji→kana) → kanaToIpa.
 * PHASE 1: native kana/katakana → IPA (kana.ts) + Sino-Japanese numbers. PHASE 2: kanji → kana via a 60k
 * whole-word reading map + per-kanji on/kun/rendaku fallback (kanji.ts), and orthographic bunsetsu
 * segmentation of spaceless text. Pitch accent (ꜜ) is applied by pitch.ts.
 */
import type { Phonemizer } from "../../registry.ts";
import { makeSymbolNormalizer } from "../../core/normalizeSymbols.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { kanaToIpa, kanaToMorae, segmentsToMorae } from "./kana.ts";
import { applyReadingSegments, applyReadings, segmentText, headsCompound } from "./kanji.ts";
import { numberToKana } from "./numbers.ts";
import { readCounter } from "./counters.ts";
import { normalizeJapanese } from "./normalize.ts";

// Fold hiragana → katakana (kanaToIpa treats them identically). Counter readings are injected as katakana so
// segmentText's hiragana-specific は→わ / を particle heuristic can't corrupt an internal は/へ (2泊→にはく → にわく).
const toKatakana = (s: string): string =>
    s.replace(/[ぁ-ゖ]/gu, (c) => String.fromCodePoint(c.codePointAt(0)! + 0x60));
import { accentNucleus, placeDownstep } from "./pitch.ts";
import { MANIFEST } from "./manifest.ts";
import { rewrite } from "../../core/provenance.ts";

// Japanese clause punctuation → canonical pause marks (from japanese.jsonc).
const CLAUSE_MARK = MANIFEST.clausePunctuation;
// A Japanese-script run (kanji incl. Ext-A/B + iteration marks + hiragana + katakana + long mark), a digit
// run, or clause punctuation. Bunsetsu spaces inserted by segmentText split runs into phrase-sized tokens.
const TOKEN =
    /([㐀-鿿\u{20000}-\u{2a6df}々〻ぁ-ゖァ-ヺー゛゜]+)|(\d+)|([。．.！!？?、，,])/gu;
const KANA_ONLY = /[^ぁ-ゖァ-ヺー]/gu; // strip anything the reading pass left un-converted (unresolved kanji)

// Katakana loans, read by the ordinary kana engine.
// UNITS moved to normalize.ts, which must resolve them before its decimal and exponent rules break the
// number-adjacency this tier matches on; see UNIT_KANA there. Percent stays, because nothing reorders it.
// ⚠ Without a currency declaration the sign is DROPPED outright, so "$5" and "5" read identically. Japanese
// prose normally writes 円 and ドル as words rather than using a sign, but the reading is not in doubt and a
// dropped sign is silent content loss wherever one appears.
const SYMBOLS = makeSymbolNormalizer({
    percent: MANIFEST.symbolTier.percent,
    currency: MANIFEST.symbolTier.currency,
    units: MANIFEST.symbolTier.units,
    exponentWords: MANIFEST.symbolTier.exponentWords,
    bareExponent: MANIFEST.symbolTier.bareExponent,
    unspacedScript: MANIFEST.symbolTier.unspacedScript,
    ampersand: MANIFEST.symbolTier.ampersand,
    multiply: MANIFEST.symbolTier.multiply,
});

class JapanesePhonemizer implements Phonemizer {
    text(input: string): string {
        // SYMBOLS first, because its % rule matches a NUMBER directly before the sign and
        // normalization's decimal rewrite (1.5 → 1点ゴ) removes that adjacency. Normalization then folds
        // the widths, resolves the remaining numeric surface forms, and nativizes embedded Latin so it
        // never reaches the English fallback in core/foreign.ts.
        input = normalizeJapanese(SYMBOLS(input));
        // Normalise full-width digits ０-９ → ASCII so the number path fires (３個 → さんこ, ２０２４年 → …); the \d
        // token and numberToKana are ASCII-only.
        input = rewrite(input, /[０-９]/gu, (d) => String.fromCodePoint(d.codePointAt(0)! - 0xfee0));
        // Number + counter (助数詞): fuse a digit run + following counter kanji into its euphonic kana reading
        // (1本→いっぽん, 3個→さんこ, 2024年→にせんにじゅうよねん) BEFORE segmentation, so it flows through the kana path.
        // readCounter returns null for a non-counter kanji (or out-of-range n) → the digits pass through unchanged.
        // Suppress the fusion when the counter kanji HEADS a dictionary compound (3時間, 3年生): splitting it off
        // would orphan the trailing kanji into a wrong isolated reading (間→あいだ, 生→なま). See headsCompound.
        input = rewrite(
            input,
            // ⚠ `つ` is listed EXPLICITLY beside Han: it is the one counter written in hiragana, and
            // matching it is what lets 1つ reach readCounter at all. Widening this to kana generally
            // would be wrong — a digit is followed by an ordinary particle constantly (3の, 5は).
            /(\d+)(\p{Script=Han}|つ)/gu,
            (m0, num, ctr, offset: number, str: string) => {
                if (headsCompound(str.slice(offset + num.length))) return m0;
                const reading = readCounter(Number(num), ctr);
                return reading === null ? m0 : toKatakana(reading);
            },
        );
        // segmentText inserts bunsetsu spaces first, then assembleClauses runs the standard clause skeleton.
        return assembleClauses(segmentText(input), TOKEN, (m, sink) => {
            if (m[1]) {
                const segments = readingSegments(m[1]); // kanji → kana per morpheme (boundaries kept)
                const reading = segments.join("");
                const morae = segmentsToMorae(segments);
                if (morae)
                    sink.emit(
                        placeDownstep(morae, accentNucleus(m[1], reading)),
                    ); // pitch: surface m[1] disambiguates
            } else if (m[2]) {
                const ipa = kanaToIpa(numberToKana(Number(m[2]), m[2]));
                if (ipa) sink.emit(ipa);
            } else if (m[3]) {
                const mk = CLAUSE_MARK[m[3]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** One Japanese word/token → canonical IPA (kanji readings + pitch downstep, so kanji tokens work too). */
export function phonemizeWord(word: string): string {
    const segments = readingSegments(word);
    const reading = segments.join("");
    const morae = segmentsToMorae(segments);
    return morae === null
        ? ""
        : placeDownstep(morae, accentNucleus(word, reading));
}

/** Reading segments for a word, with the unresolvable tail dropped from each.
 *
 *  No exception list is needed for compounds that legitimately DO coalesce across the boundary: those are
 *  exactly the ones whose stored reading is NOT the sum of their characters' readings, so
 *  `alignCompoundReading` finds no split and leaves them as a single segment. 小売 こうり stays koːri because
 *  売 has no reading うり; 子牛 こうし splits こ|うし because both parts are attested readings. The mechanism
 *  decides it from the data instead of a hand-maintained list. */
function readingSegments(word: string): string[] {
    return applyReadingSegments(word)
        .map((s) => s.replace(KANA_ONLY, ""))
        .filter((s) => s !== "");
}

/** One Japanese word/token → canonical IPA, SEGMENTAL only (no pitch downstep) — for segmental validation. */
export function phonemizeWordSegmental(word: string): string {
    return kanaToIpa(applyReadings(word).replace(KANA_ONLY, "")) ?? "";
}

/** Build the Japanese phonemizer (kana + numbers + kanji readings + bunsetsu segmentation). */
export function createJapanese(): Phonemizer {
    return new JapanesePhonemizer();
}

File: csharp/Vernacula.Phonemizer/Languages/Japanese/Japanese.cs
/**
 * Japanese (ja) phonemizer — Standard/Tokyo, canonical IPA.
 * Ported from src/languages/japanese/japanese.ts — see that file for the corpus evidence.
 */
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Japanese;

public static class JapanesePhonemizer
{
    private static readonly JsRe HIRAGANA_RANGE = JsRegex.Compile("[ぁ-ゖ]", "gu");

    // Counter readings are injected as KATAKANA so segmentText's hiragana-only は→わ / を particle heuristic
    // cannot corrupt an internal は/へ (2泊 → にはく → *にわく).
    private static string ToKatakana(string s) =>
        HIRAGANA_RANGE.Replace(s, c => char.ConvertFromUtf32(Js.CodePointAt0(c.Value) + 0x60));

    private static IReadOnlyDictionary<string, string> CLAUSE_MARK => Manifest.MANIFEST.ClausePunctuation;
    private static readonly JsRe TOKEN =
        JsRegex.Compile("([㐀-鿿\\u{20000}-\\u{2a6df}々〻ぁ-ゖァ-ヺー゛゜]+)|(\\d+)|([。．.！!？?、，,])", "gu");
    private static readonly JsRe KANA_ONLY = JsRegex.Compile("[^ぁ-ゖァ-ヺー]", "gu"); // strip anything the reading pass left un-converted (unresolved kanji)

    private static readonly Func<string, string> SYMBOLS = NormalizeSymbols.MakeSymbolNormalizer(new SymbolData
    {
        Percent = Manifest.MANIFEST.SymbolTier.Percent,
        Currency = Manifest.MANIFEST.SymbolTier.Currency,
        Units = Manifest.MANIFEST.SymbolTier.Units,
        ExponentWords = Manifest.MANIFEST.SymbolTier.ExponentWords,
        BareExponent = Manifest.MANIFEST.SymbolTier.BareExponent,
        Ampersand = Manifest.MANIFEST.SymbolTier.Ampersand,
        Multiply = Manifest.MANIFEST.SymbolTier.Multiply,
        UnspacedScript = Manifest.MANIFEST.SymbolTier.UnspacedScript,
    });

    private static readonly JsRe FULLWIDTH_DIGIT = JsRegex.Compile("[０-９]", "gu");
    private static readonly JsRe COUNTER_FUSION = JsRegex.Compile("(\\d+)(\\p{Script=Han}|つ)", "gu");

    private sealed class Engine : ILanguage
    {
        public string Text(string input)
        {
            // SYMBOLS must run before normalization: its % rule matches a number directly before the sign, and
            // the decimal rewrite (1.5 → 1点ゴ) destroys that adjacency.
            input = Normalize.NormalizeJapanese(SYMBOLS(input));
            input = FULLWIDTH_DIGIT.Replace(input, d => char.ConvertFromUtf32(Js.CodePointAt0(d.Value) - 0xfee0));
            // Counter fusion runs BEFORE segmentation so the fused reading flows through the kana path. The
            // headsCompound guard suppresses it when the counter kanji heads a dictionary compound (3時間,
            // 3年生); splitting there would orphan the trailing kanji into a wrong isolated reading.
            var src = input;
            input = COUNTER_FUSION.Replace(src, m =>
            {
                var num = m.Groups[1].Value;
                var ctr = m.Groups[2].Value;
                if (Kanji.HeadsCompound(src[(m.Index + num.Length)..])) return m.Value;
                var reading = Counters.ReadCounter(Js.Number(num), ctr);
                return reading is null ? m.Value : ToKatakana(reading);
            });
            return Clauses.AssembleClauses(Kanji.SegmentText(input), TOKEN, (m, sink) =>
            {
                if (m.Groups[1].Success && m.Groups[1].Value.Length > 0)
                {
                    var segments = ReadingSegments(m.Groups[1].Value); // kanji → kana per morpheme (boundaries kept)
                    var reading = string.Concat(segments);
                    var morae = Kana.SegmentsToMorae(segments);
                    if (morae is not null)
                        sink.Emit(Pitch.PlaceDownstep(morae, Pitch.AccentNucleus(m.Groups[1].Value, reading))); // pitch: surface m[1] disambiguates
                }
                else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0)
                {
                    var ipa = Kana.KanaToIpa(Numbers.NumberToKana(Js.Number(m.Groups[2].Value)));
                    if (!string.IsNullOrEmpty(ipa)) sink.Emit(ipa);
                }
                else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
                {
                    var mk = CLAUSE_MARK.GetValueOrDefault(m.Groups[3].Value);
                    if (mk is not null) sink.Pause(mk);
                }
            });
        }
    }

    /**
     * One Japanese word/token → canonical IPA (kanji readings + pitch downstep, so kanji tokens work too).
     */
    public static string PhonemizeWord(string word)
    {
        var segments = ReadingSegments(word);
        var reading = string.Concat(segments);
        var morae = Kana.SegmentsToMorae(segments);
        return morae is null ? "" : Pitch.PlaceDownstep(morae, Pitch.AccentNucleus(word, reading));
    }

    /** Reading segments for a word, with the unresolvable tail dropped from each. */
    private static List<string> ReadingSegments(string word) =>
        Kanji.ApplyReadingSegments(word)
            .Select(s => KANA_ONLY.Replace(s, ""))
            .Where(s => s != "")
            .ToList();

    /**
     * One Japanese word/token → canonical IPA, SEGMENTAL only (no pitch downstep) — for segmental validation.
     */
    public static string PhonemizeWordSegmental(string word) =>
        Kana.KanaToIpa(KANA_ONLY.Replace(Kanji.ApplyReadings(word), "")) ?? "";

    /** Build the Japanese phonemizer (kana + numbers + kanji readings + bunsetsu segmentation). */
    public static ILanguage CreateJapanese() => new Engine();

    internal static void RegisterSelf() => Registry.Register("japanese", () => CreateJapanese());
}
